src/charm.py

```diff
@@ -152,12 +141,16 @@ def _configure(self, _):
                 logger.error(f"Failed to replan; pebble plan: {container.get_plan().to_dict()}")
                 raise e
 
-        if (
-            isinstance(self.unit.status, BlockedStatus)
-            and self.unit.status.message != CORRUPT_PROMETHEUS_CONFIG_MESSAGE
-        ):
+        # Now reload config to make sure it is valid
+        external_url = urlparse(self._external_url)
+        prometheus_server = Prometheus(web_route_prefix=external_url.path)
+        reloaded = prometheus_server.reload_configuration()
+        if not reloaded:
+            logger.error("Prometheus failed to reload the configuration")
+            self.unit.status = BlockedStatus(CORRUPT_PROMETHEUS_CONFIG_MESSAGE)
             return
 
+        logger.info("Prometheus configuration reloaded")
         self.unit.status = ActiveStatus()
```

Comment:
This changes the behavior significantly. Previously, the only blocked status we would "auto-heal" from was the Prometheus config being corrupt. Now, as long as the config is not corrupt, any other reason for blocks would also be discarded. Is this intentional?

Reply:
Further, why is a subsequent reload needed after a replan? Replan will restart all services which changed layers. To my knowledge, the reason why we had a conditional check deciding whether to replan or reload, is to avoid unnecessary downtime in the case where a reload is sufficient.

I first thought that it might be because you wanted a trigger for detecting invalid configs, but as far as I've been told (thanks @balbirthomas for that clarification), Prometheus will exit with an error status if you try to launch it with a broken config?

Reply:
> except the rare occasion when layer changes

This is the case I want to guard for, as it would otherwise mean it first is restarted, and then additionally reloaded

Reply:
Done.